readme.md
HTTP API, PHP HTTP client and webservice framework
==================================================
HTTP API is a PHP HTTP client that makes it easy to send HTTP requests and
trivial to integrate with web services.

- Manages things like persistent connections, represents query strings as
  collections, simplifies sending streaming POST requests with fields

- Can send both synchronous and asynchronous requests using the same interface
  without requiring a dependency on a specific event loop.

### Installing via Composer

```bash
"repositories": [
    {
        "type": "vcs",
        "url": "http://git.tnq.co.in/common/http-api.git"
    },
],
"require": {
    "php": ">=5.4.0",
    "tnq/httpapi": "latest tag"
},
```

### Make a Request

You can send requests with HttpApi using a Tnq\HttpApi\HttpRequest object.

```php
use Tnq\HttpApi\HttpRequest;

$httpApi = new HttpRequest();
$httpApi->setUrl("URL"); /* Request url to process */
$request = $httpApi->build();
$response = $request->execute();
```
### Overwrite default params before sending http reqeust

##### Set Reqeust Method

By default Reqeust method for HTTP API in GET method, to overwrite the default method type for the request as follows

```php
$httpApi->setMethod("GET"); //"POST", "GET", "PUT", "DELETE"
```
##### Set Authentication

To access the Authentication based URL need to sent authentication method as TRUE by default authentication method is open for every reqeust

```php
$httpApi->setAuth(true);
$httpApi->setAuthUsername("username");
$httpApi->setAuthPassword("password");
$httpApi->setAuthType("authentication type"); //"Basice", "Digest"
```
##### Set Headers

While sending the request we can add Header detail with the http reqeust

```php
$httpApi->setHeaders(
    [
        'X-Foo-Header' => 'value',
        'content-type' => 'application/json'
    ]
);
```
##### Set Retry

If the reqeust get faild there is a option to retry the reqeust, by default retry is set to false and there is option to set number of retry to be call.

```php
$httpApi->setRetry(true);
$httpApi->setRetryCount(5);
```

We can specify type of error to be retry using status codes

```php
$httpApi->setTypeOfErrorToRetry([500]);
```
##### Set Delay

we can set delay between the reqeust, by default delay set to 1

```php
$httpApi->setDelay(1000); /*milli secound*/
```

##### Set Data

You can send data as parameters for both GET and POST by following method

```php
$httpApi->setData("array of parameters");
```

If you want to send data as raw data you can send by following method

```php
$httpApi->setDataAsBody(true);
$httpApi->setData("Raw body content");
```

### Using Responses

In the previous examples, we retrieved a ```$response``` variable. This value is actually a Tnq\HttpApi\ResponsetHandler object and contains lots of helpful information.

##### Status code

To get the status code of the request

```php
$response->getStatusCode(); /*200 response*/
```

##### Body Content

The body of a response can be retrieved and cast to a string.

```php
$response->getBodyContent(); /*Body Content*/
```

##### JSON Responses

You can more easily work with JSON responses using the getJsonContent() method of a response.

```php
$response->getJsonContent(); /*Returns JSON formate body contents as array*/
```
##### XML Responses

You can use a responseâ€™s getXmlContent() method to more easily work with responses that contain XML data.

```php
$response->getXmlContent(); /*Returns XML formate body contents as array*/
```
##### Header Details

You can get the whole header details as an array using getHeaders().

```php
$response->getHeaders(); /*Returns header details as array*/
```
To get a specific formate of header details by following method

###### As String

```php
$response->getHeaderAsString('content-type');
/*return header details as string formate OUTPUT:application/json */
```
###### As Array

```php
$response->getHeaderAsArray('content-type');
#return header details as string formate OUTPUT:["0"=> "application/json"]
```

If you want to get detail other then mentioned here you can directly get it from $response object

### Exception handling

You will get two kind of exception from HTTPAPI reqeust

###### Request exception

Throws exception while is there any details given from the user is not valid, you can overwrite the exception using ``` use Tnq\HttpApi\Exception\RequestException;```

###### Response exception

Throws exception while executing http reqeust, you can overwrite the exception using
``` use Tnq\HttpApi\Exception\ResponseException;```


changelog.MD
## [1.0.0] - 2015-07-29
### Added
- Inital Version of Http Router.
- Front Controller Pattern

## [1.0.1] - 2015-09-04
### Added
- Allow cors with multiple domain.

## [1.0.2] - 2016-08-24
### Changed
- Failure Response get error code from exception throw from wrapper.

## [1.0.3] - 2016-12-29
### Changed
- Revert symfony validation.

## [1.0.4] - 2017-02-20
### Changed
- Error response returns as html format.

## [1.0.5] - 2017-08-02
### Added
- getSilex function Added.

## [1.0.6] - 2018-05-10
### Added
- Options added for error response return as JSON.
- Logger interface Added.
- Throw JSON error message, if json_encode fails.


app/Tnq/Routes/Response/HttpResponse.php
<?php
namespace Tnq\Routes\Response;

use Symfony\Component\HttpFoundation\Response;

class HttpResponse {
    private $headers;
    private $content;

    public function __construct(
        Response $response,
        ResponseHandlerComponent $responseHandler
    ) {
        $this->responseHandler = $responseHandler;
        $this->response = $response;
    }

    public function addHeader($name, $value) {
        $this->response->headers->set($name, $value);
    }

    public function setContent($content) {
        $this->response->setContent($content);
    }

    private function defaultResponse() {
        $defaultResponseData = [
            'status' => true,
            'message' => null,
            'data' => null
        ];

        return $defaultResponseData;
    }

    public function generateResponse(array $responseData) {
        $defaultResponseData = $this->defaultResponse();
        $filteredResponse = array_replace_recursive(
            $defaultResponseData, $responseData
        );

        $responsecontent  = $this->responseHandler->generateResponse(
            $filteredResponse
        );
        $this->response->setContent($responsecontent);
        $this->addHeader(
            'Content-Type',
            $this->responseHandler->getDefaultContentType()
        );

        return $this->response;
    }

    private function setHeaderCookies($response, $cookies) {
        foreach ($cookies as $cookie) {
            $response->headers->setCookie(new Cookie($cookie[0], $cookie[1]));
        }
    }

    private function clearCookies($response, $cookie) {
        $response->headers->clearCookie($cookie);
    }
}


app/Tnq/Routes/Response/ResponseFactory.php
<?php
namespace Tnq\Routes\Response;

use Symfony\Component\Serializer\Serializer;
use Symfony\Component\Serializer\Encoder\JsonEncoder;
use Symfony\Component\Serializer\Normalizer\GetSetMethodNormalizer;
use Symfony\Component\HttpFoundation\Response;
use Tnq\Routes\Response\ResponseHandlerComponent;

class ResponseFactory {
    private $cors;

    public function __construct($cors = []) {
        $this->cors = $cors;
    }


    public function createHandler($type = 'application/json') {
        $encoders = [new JsonEncoder()];
        $normalizers = [new GetSetMethodNormalizer()];
        $serializer =  new Serializer($normalizers, $encoders);
        $responseHandler = new ResponseHandlerComponent(
            $serializer, $type, $this->cors
        );
        return $responseHandler;
    }
}


app/Tnq/Routes/Response/ResponseHandlerComponent.php
<?php
namespace Tnq\Routes\Response;

use Pimple;
use Symfony\Component\HttpFoundation\Cookie; //FIXME
use Doctrine\Common\Inflector\Inflector;
use Symfony\Component\HttpFoundation\BinaryFileResponse;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Serializer\Serializer;

class ResponseHandlerComponent {

    private $context;
    private $data;
    private $status;
    private $headers;
    private $requestType;
    private $serializer;
    private $responseFactory;

    public function __construct(
        Serializer $serializer, $contentType, $cors = []
    ) {
        $this->serializer = $serializer;
        $this->headers['Content-Type'] = $contentType;
        $this->cors = $cors;
    }

    public function generateResponse($responseData) {
        $result = [];
        $contentType = $this->headers['Content-Type'];
        $responseExecutors = [
            'application/json' => 'generateJsonResponse',
            'application/xml' => 'generateXmlResponse',
            'text/html' => 'generateJsonResponse'
        ];
        if (empty($responseExecutors[$contentType])) {
            die("Content-Type '". $contentType ."'' not supported");
        }
        if (is_callable([$this, $responseExecutors[$contentType]])) {
            $result = call_user_func(
                [$this, $responseExecutors[$contentType]], $responseData
            );
        }

        return $result;
    }

    private function generateDataResponse() {
        return $this->data;
    }

    private function generateXmlResponse($result) {
        $serializer = $this->serializer;
        return $serializer->serialize($result, 'xml');
    }

    private function generateJsonResponse($result) {
        if (json_encode($result) === false) {
            throw new \Exception(
                'Response Contains Json Parse Error States: '.json_last_error_msg()
            );
        }

        return json_encode($result);
    }

    private function sendCliResponse($data) {
        return $data;
    }

    public function setCrosHeaders(Response $response) {
        $corsValues = $this->crosOriginHandler();
        $allowedOrigin = $this->getAllowedOrigin(
            $corsValues['origin']
        );
        $headers = $response->headers;
        $headers->set(
            'Access-Control-Allow-Origin', $allowedOrigin
        );
        $headers->set(
            'Access-Control-Allow-Headers', $corsValues['accept_headers']
        );
        $headers->set(
            'Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, DELETE'
        );
        $headers->set(
            'Access-Control-Allow-Credentials', $corsValues['allow_credentials']
        );
    }

    public function createResponse($data, $status=200) {
        $response = new Response(null , $status);
        $this->setCrosHeaders($response);
        $httpResponse = new HttpResponse($response, $this);
        return $httpResponse->generateResponse($data);
    }

    public function createTextResponse($data, $status=200) {
        $response = new Response($data, $status);
        $this->setCrosHeaders($response);
        return $response;
    }

    private function crosOriginHandler() {
        $keyExists = array_key_exists('cors', $this->cors);
        if ($keyExists === false) {
            return [
                'origin' => ['*'],
                'allow_credentials' => 'false',
                'accept_headers' => 'content-type, x-requested-with',
                'accept_methods' => 'GET, POST, OPTIONS, PUT, DELETE'
            ];
        }

        return $this->cors['cors'];
    }

    private function getOriginHeader() {
        if (isset($_SERVER['HTTP_ORIGIN'])) {
            return $_SERVER['HTTP_ORIGIN'];
        }
        return null;
    }

    private function getAllowedOrigin(array $allowedOriginList) {
        $originHeader = $this->getOriginHeader();
        if (empty($originHeader)) {
            return null;
        }
        $matchedKey = array_search($originHeader, $allowedOriginList);
        if ($matchedKey >= 0) {
            $matchedOrigin = $allowedOriginList[$matchedKey];
            return $matchedOrigin;
        }
        return null;
    }

    public function getDefaultContentType() {
        return $this->headers['Content-Type'];
    }
}


app/Tnq/Routes/Request/FrontController.php
<?php
namespace Tnq\Routes\Request;

use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Silex\Application;
use Tnq\Routes\Response\ResponseHandlerComponent;
use Tnq\Routes\Logger\Logger;

class FrontController {
    private $app = null;
    private $context = null;

    public function __construct(
        $context, array $routes, ResponseHandlerComponent $responseHandler,
        $options = []
    ) {
        $this->app = new Application();
        $this->app['debug'] = true;
        $this->context = $context;
        $this->responseHandler = $responseHandler;
        $this->setupRoutes($routes, $options);
    }

    private function getError($message, $status = false) {
        $errorData = [];
        $errorData['data'] = " ";
        $errorData['message'] = $message;
        $errorData['status'] = $status;
        return $errorData;
    }

    public function setupRoutes(array $routes, array $options = []) {

        foreach ($routes as $route) {
            $s = $route['invoke'];
            $pos = strpos($s, '->');
            $className = trim(substr($s, 0, $pos));
            $method = trim(substr($s, $pos + strlen('->')));
            $this->app->match(
                $route['url'], function(Request $request
            ) use($className, $method) {
                return call_user_func(
                    [$this->context[$className], $method], $request
                );
            })->method($route['method']);
        }

        $this->app->match("{url}", function($url) {
            return $this->responseHandler->createTextResponse('OK', 200);
        })->assert('url', '.*')->method('OPTIONS');

        $this->app->error(function (\Exception $e, $code) use ($options) {
            $error = $e->getMessage();
            if (array_key_exists('errorLogger', $options) === true) {
                $logger = $options['errorLogger'];
                if ($logger instanceof Logger === false) {
                    throw new \Exception(
                        'logger object should implements the interface Tnq\Routes\Logger\Logger'
                    );
                }

                $logger->logException($e);
            }

            if ((array_key_exists('errorAsJson', $options) === true) &&
                ($options['errorAsJson'] === true)
            ) {
                $code = $e->getCode() ? $e->getCode() : 500;
                $response = $this->responseHandler->createResponse(
                    $this->getError($error),
                    $code
                );

                return $response;
            }

            $response = $this->responseHandler->createTextResponse($error, 500);
            $response->headers->set('Content-Type', 'text/html');
            //$response->headers->set('X-Status-Code', 500);

            return $response;
        });
    }

    public function getSilex() {
        return $this->app;
    }

    public function run() {
        $this->app->run();
    }
}
